File: 1305-number-of-visible-people-in-a-queue/number-of-visible-people-in-a-queue.java
class Solution {
    public int[] canSeePersonsCount(int[] heights) {
        Deque<Integer> stack = new ArrayDeque<>();
        int[] numPeople = new int[heights.length];
        for(int i = heights.length - 1; i >= 0; i--){
            int count = 0;
            while(!stack.isEmpty() && stack.peek() < heights[i]){
                stack.pop();
                count++;
            }

            numPeople[i] = stack.isEmpty() ? count : count + 1;

            stack.push(heights[i]);
        }
        return numPeople;
    }
}

File: 1305-number-of-visible-people-in-a-queue/number-of-visible-people-in-a-queue.cpp
class Solution {
public:
    vector<int> canSeePersonsCount(vector<int>& heights) {
        int n = heights.size();
        vector<int> numPeople(n, 0);
        stack<int> stack;
        for (int i = n - 1; i >= 0; --i) {
            int count = 0;
            while (!stack.empty() && stack.top() < heights[i]) {
                stack.pop();
                ++count;
            }
            numPeople[i] = stack.empty() ? count : count + 1;
            stack.push(heights[i]);
        }
        return numPeople;
    }
};
